torch/csrc/jit/codegen/cuda/lower_warp_reduce.cpp

```diff
@@ -0,0 +1,546 @@
+#include <torch/csrc/jit/codegen/cuda/expr_evaluator.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+#include <torch/csrc/jit/codegen/cuda/lower_warp_reduce.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+//! A simple DCE for eliminating the
+//!  parallel broadcasts that has been fused
+//!  and their corresponding allocates
+class eliminateDeadBroadcastAndAllocate {
+ public:
+  static std::vector<kir::Expr*> run(const std::vector<kir::Expr*>& exprs) {
+    eliminateDeadBroadcastAndAllocate dce(exprs);
+    return dce.result_exprs_;
+  }
+
+ private:
+  eliminateDeadBroadcastAndAllocate(const std::vector<kir::Expr*>& exprs)
+      : ir_builder_(GpuLower::current()->kernel()) {
+    findLiveTvs(exprs);
+    findDeadTvs();
+    eliminateDeadCode(exprs);
+  }
+
+  void findLiveTvs(const std::vector<kir::Expr*>& exprs) {
+    for (auto expr : exprs) {
+      if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+        findLiveTvs(for_loop->body().exprs());
+        continue;
+      } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+        findLiveTvs(ite->thenBody().exprs());
+        findLiveTvs(ite->elseBody().exprs());
+        continue;
+      }
+
+      if (auto allocate = dynamic_cast<kir::Allocate*>(expr)) {
+        if (allocate->memoryType() == MemoryType::Local) {
+          if (auto kir_tv =
+                  dynamic_cast<kir::TensorView*>(allocate->buffer())) {
+            // We know only tvs that we'd want to consider are broadcast outputs
+            if (kir_tv->fuserTv()->definition()->isA<BroadcastOp>()) {
+              candidate_tvs_.push_back(kir_tv);
+              candidate_tv_set_.insert(kir_tv);
```

Comment:
This is in case a deterministic loop order is needed. It's not absolutely necessary for now and could remove if we really want to .

Reply:
So, the only place I see that uses `candidate_tvs_` is line 67. It is used to populate `dead_tvs_`. Since that's also an unordered set, I don't see why ordering of candidate tvs could matter. Am I missing something?

Reply:
No that's all. I guess it shouldn't matter with the ordering. I will remove it and maybe re-factor if we really need deterministic ordering on this.